internal/pkg/docker/orchestrator/orchestrator.go

```diff
@@ -452,6 +485,80 @@ func (o *Orchestrator) waitForContainerToStart(ctx context.Context, id string) e
 	}
 }
 
+func (o *Orchestrator) waitForContainerDependencies(ctx context.Context, name string, a *runTaskAction) error {
+	var deps []string
+	for depName, state := range a.task.Containers[name].DependsOn {
+		deps = append(deps, fmt.Sprintf("%s->%s", depName, state))
+	}
+	logMsg := strings.Join(deps, ", ")
+	fmt.Printf("Waiting for container %q dependencies: [%s]\n", name, color.Emphasize(logMsg))
+	eg, ctx := errgroup.WithContext(ctx)
+	for name, state := range a.task.Containers[name].DependsOn {
+		name, state := name, state
+		eg.Go(func() error {
+			ctrId := o.containerID(name)
+			isEssential := a.task.Containers[name].IsEssential
+			ticker := time.NewTicker(700 * time.Millisecond)
+			defer ticker.Stop()
+			for {
+				select {
+				case <-ticker.C:
+				case <-ctx.Done():
+					return ctx.Err()
+				}
+				switch state {
+				case ctrStateStart:
+					err := o.waitForContainerToStart(ctx, ctrId, isEssential)
+					if err != nil {
+						return err
+					}
+					if isEssential {
+						log.Successf("Successfully started container %s\n", ctrId)
+						return nil
+					}
+					fmt.Printf("non-essential container %q started successfully\n", ctrId)
+					return nil
+				case ctrStateHealthy:
+					healthy, err := o.docker.IsContainerHealthy(ctx, ctrId)
```

Comment:
can the container be "starting"? should we _wait_ for it to be healthy?

Reply:
This is what exactly the current implementation is doing
In docker package `IsContainerHealthy` returns `false` and `nil` if container is `starting`/ still not reached `healthy`.

Docker tries to update the state of container based on number of retries in Healthcheck command 
https://github.com/aws/copilot-cli/blob/mainline/internal/pkg/docker/dockerengine/dockerengine.go#L396:L412

```
HEALTHCHECK --interval=1m --timeout=30s --start-period=5s --retries=3 CMD echo "Hello"
```

so in `orchestrator` pkg, error out only if the health check state is `unhealthy`/`container exited` else we check the status every `700` milliseconds until container reaches `healthy`.